src/github.com/couchbase/sync_gateway/base/sharded_sequence_clock.go

```diff
@@ -0,0 +1,294 @@
+	shardedClockExpvars = expvar.NewMap("syncGateway_index_clocks")
+}
+
+const (
+	kIndexPartitionKey       = "_idxPartitionMap"
+	kIndexPrefix             = "_idx"
+	kCountKeyFormat          = "_idx_c:%s:count"    // key
+	kClockPartitionKeyFormat = "_idx_c:%s:clock-%d" // key, partition index
+)
+
+type PartitionStorage struct {
+	Uuid  string   `json:"uuid"`
+	Index uint16   `json:"index"`
+	VbNos []uint16 `json:"vbNos"`
+}
+
+type IndexPartitionMap map[uint16]uint16 // Maps vbuckets to index partition value
+
+type IndexPartitions struct {
+	VbMap         IndexPartitionMap
+	PartitionDefs []PartitionStorage
+}
+
+func NewIndexPartitions(partitions []PartitionStorage) *IndexPartitions {
+
+	indexPartitions := &IndexPartitions{
+		PartitionDefs: make([]PartitionStorage, len(partitions)),
+	}
+	copy(indexPartitions.PartitionDefs, partitions)
+
+	indexPartitions.VbMap = make(map[uint16]uint16)
+	for _, partition := range partitions {
+		for _, vbNo := range partition.VbNos {
+			indexPartitions.VbMap[vbNo] = partition.Index
+		}
+	}
+	return indexPartitions
+}
+
+// Index partitionsfor testing
+func TestIndexPartitions() *IndexPartitions {
+
+	maxVbNo := uint16(1024)
+	partitions := make(IndexPartitionMap, maxVbNo)
+
+	numPartitions := uint16(32)
+	partitionDefs := make([]PartitionStorage, numPartitions)
+	vbPerPartition := maxVbNo / numPartitions
+	for partition := uint16(0); partition < numPartitions; partition++ {
+		storage := PartitionStorage{
+			Index: partition,
+			VbNos: make([]uint16, vbPerPartition),
+		}
+		for index := uint16(0); index < vbPerPartition; index++ {
+			vb := partition*vbPerPartition + index
+			partitions[vb] = partition
+			storage.VbNos = append(storage.VbNos, vb)
+		}
+		partitionDefs[partition] = storage
+	}
+	return &IndexPartitions{
+		VbMap:         partitions,
+		PartitionDefs: partitionDefs,
+	}
+}
+
+// Index partition map for testing
+func TestIndexPartitionMap() IndexPartitionMap {
+	return TestIndexPartitions().VbMap
+}
+
+// ShardedClock maintains the collection of clock shards (ShardedClockPartitions), and also manages
+// the counter for the clock.
+type ShardedClock struct {
+	baseKey       string                            // key prefix used to build keys for clock component docs
+	counter       uint64                            // count value for clock
+	countKey      string                            // key used to incr count value
+	partitionMap  *IndexPartitions                  // Index partition map
+	partitions    map[uint16]*ShardedClockPartition // Clock partitions - one doc written per partition
+	bucket        Bucket                            // Bucket used to store clocks
+	partitionKeys []string                          // Keys for all partitions.  Convenience to avoid rebuilding set from partitions
+}
+
+// ShardedClockPartition manages storage for one clock partition, where a clock partition is a set of
+// {vb, seq} values for a subset of vbuckets.
+type ShardedClockPartition struct {
```

Comment:
I agree that this is problematic.  I think the rest of the distributed index code (kv_change_index, kv_channel_index, etc) uses 'partition' to describe the same concept as 'pindex'.  

I'm not a huge fan of the term 'pindex' - I don't think "a subset of vbuckets" when I hear that - but I agree that the overlapping terminology between partition, shard, and pindex is only going to cause us pain down the road.  Let's leave it out of this PR, but plan to talk about how to refactor to consistent usage.